src/protocol/clock.rs

```diff
@@ -0,0 +1,227 @@
+use std::cell::RefCell;
+pub use std::time::Duration;
+use std::time::SystemTime;
+
+use thiserror::Error;
+
+use crate::static_time;
+
+pub type SinceUnixEpoch = Duration;
+
+#[derive(Error, Debug)]
+pub enum ClockError {
+    #[error("overflow error")]
+    OverflowError,
+}
+
+pub trait Time: Sized {
+    fn now() -> SinceUnixEpoch;
+
+    fn now_add(duration: &Duration) -> Option<SinceUnixEpoch> {
+        Self::now().checked_add(*duration)
+    }
+    fn now_sub(duration: &Duration) -> Option<SinceUnixEpoch> {
+        Self::now().checked_sub(*duration)
+    }
+    fn now_periods(period: &Duration) -> Option<u128> {
+        Self::now().as_nanos().checked_div(period.as_nanos())
+    }
+    fn now_add_periods(duration: &Duration, period: &Duration) -> Option<u128> {
+        match Self::now_add(duration) {
+            None => None,
+            Some(time) => time.as_nanos().checked_div(period.as_nanos()),
+        }
+    }
+    fn now_sub_periods(duration: &Duration, period: &Duration) -> Option<u128> {
+        match Self::now_sub(duration) {
+            None => None,
+            Some(time) => time.as_nanos().checked_div(period.as_nanos()),
+        }
+    }
+}
```

Comment:
@da2ce7 Somehow, I would like to eliminate the prefix `now_` in all methods. But I do not know if it makes sense:

```rust
pub trait Time: Sized {
    fn now() -> SinceUnixEpoch;
}

pub trait TimeNow: Sized {
    fn add(duration: &Duration) -> Option<SinceUnixEpoch> {
        Time::now().checked_add(*duration)
    }
    fn sub(duration: &Duration) -> Option<SinceUnixEpoch> {
        Time::now().checked_sub(*duration)
    }
    fn periods(period: &Duration) -> Option<u128> {
        Time::now().as_nanos().checked_div(period.as_nanos())
    }
    fn add_periods(duration: &Duration, period: &Duration) -> Option<u128> {
        match Self::add(duration) {
            None => None,
            Some(time) => time.as_nanos().checked_div(period.as_nanos()),
        }
    }
    fn sub_periods(duration: &Duration, period: &Duration) -> Option<u128> {
        match Self::sub(duration) {
            None => None,
            Some(time) => time.as_nanos().checked_div(period.as_nanos()),
        }
    }
}
```

`TimeNow` could also be `CurrentTime`.

Reply:
@josecelano Thanks, and done!